### PR TITLE
Added NodeRecycleThread in SeLion Grid which will recycle the node on…

### PR DESCRIPTION
**Pull request by sundaramrajendran**
…ce all sessions are completed

### CODE REVIEW

File: server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java [hunk 1]
```diff
@@ -166,10 +166,7 @@ public boolean release(String downloadJSON) {
 
         synchronized (this) {
             forceShutDown = true;
-            shutDownInProgress = true;
-            if (getTotalUsed() <= 1) {
-                shutdownNode();
-            }
+            startNodeRecycleThread();
```
Comment (sebady) [inline server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java:170]: Minor but isn't there a possibility of two threads starting here on release upgrade cycle?

File: server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java [hunk 2]
```diff
@@ -249,4 +240,23 @@ public synchronized void shutdownNode() {
...
 
+    /**
+     * Thread will recycle the node when all active sessions are completed
+     */
+    class NodeRecycleThread extends Thread {
+
+        @Override
+        public void run() {
+            appendMsgToCustomLog("Started NodeRecycleThread to restart the node once all the sessions are completed.");
+            while (getTotalUsed() > 1) {
```
Comment (sebady) [inline server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java]: Don't we want to wait till getTotalUsed is >0 instead?  I.e let the "last" session timeout?

Comment (sundaramrajendran) [inline server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java]: @sebady I agree your comment, we have to wait till last session timeout.

File: server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java (hunk 1 above)
Comment (sundaramrajendran) [inline server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java:170]: @sebady May be, I will fix it

**Comment by sundaramrajendran** [comment]
@sebady I have addressed all the review comments

**Comment by mach6** [comment]
@sebady looks like this is merged already..?

File: server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java [hunk 3]
```diff
@@ -212,17 +216,6 @@ public void afterSession(TestSession session) {
                 appendMsgToCustomLog("Completed session #" + totalSessionsCompleted);
             }
             appendMsgToCustomLog("Total Number of slots used : " + getTotalUsed() + " on node :" + getId());
-            if (shutDownInProgress) {
-                appendMsgToCustomLog("Proceeding with shutdown of node " + getId());
-                if (getTotalUsed() <= 1) {
-                    shutdownNode();
-                    if (forceShutDown) {
```
Comment (mach6) [inline server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java:219]: @sundaramrajendran @sebady Does force node shutdown (from the SeLion Grid Console) still work after this change?

Comment (sebady) [inline server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java:219]: The admin shutdown (via /grid/admin/GridForceRestartDelegateServlet) path did not change with this PR and should be still be fully functional.   I have tested this just now on mac.

This is now merged.

**Comment by sebady** [comment]
This is merged.

**Pull request closed without merging**